pkg/git/changelog.go

```diff
@@ -0,0 +1,267 @@
+var patchTypes = map[string]bool{"fix": true, "docs": true, "style": true, "refactor": true}
+
+func (g GitWrapper) ChangeLog(logPath string, command string, svc issues.IssueService, options GitChangeLogOptions) (GitChangeLog, error) {
+	args := append([]string{"-C", g.dir, "log"}, logPath, command)
+	out, err := pkg.NewCommand("git", args...).RunOut()
+	commits := regexp.MustCompile(`commit .{40}`).Split(out, -1)
+	repo := GetRepo(g.dir)
+	owner := GetOwner(g.dir)
+	gitChangesHolder := make([]GitChange, len(commits)-1)
+
+	for index, commit := range commits[1:] {
+		lines := regexp.MustCompile(`\n`).Split(commit, -1)
+		if len(lines) < 5 {
+			continue
+		}
+
+		title := GetCommitTitle(lines)
+		body := GetBody(lines)
+
+		if !CommitIsValid(title) {
+			change := GitChange{
+				Body:           body,
+			}
+			gitChangesHolder[index] = change
+			continue
+		}
+
+		committer := GetCommitter(lines)
+		commitType := GetCommitType(title)
+		breakingChangeFound := CommitHasBreakingChange(body)
+		issueNum := GetIssueNum(body)
+		issueLink := GetIssueLink(g.dir, issueNum)
+		storyLink, _ := GetStoryLink(svc, owner, repo, issueNum)
+
+		change := GitChange{
+			Valid:          true,
+			Title:          title,
+			Body:           body,
+			Issue:          issueNum,
+			IssueLink: 		issueLink,
+			StoryLink:      storyLink,
+			Committer:      committer,
+			CommitType:     commitType,
+			BreakingChange: breakingChangeFound,
+		}
+
+		gitChangesHolder[index] = change
+
+	}
+
+	changes := GitChanges(gitChangesHolder)
+	versionBump := changes.GetVersionBump()
+	changeLog := GitChangeLog{
+		VersionBump: versionBump,
+		Changes:     changes,
+		OutputMessage: GetChangeLogOutputMessage(versionBump, changes, options),
+	}
+
+	return changeLog, err
+}
+
+func CommitIsValid(title string) bool {
+	valid, _ := regexp.MatchString(`.*(\()(.)*(\))(:){1}`, title)
+	return valid
+}
+
+func CommitHasBreakingChange(body string) bool {
+	var breakingChangeFound = true
+	breakingChangeFormat := regexp.MustCompile(`BREAKING CHANGE: .*\n`)
+	breakingChange := breakingChangeFormat.FindAllString(body, -1)
+
+	if len(breakingChange) == 0 {
+		breakingChangeFound = false
+	}
+	return breakingChangeFound
+}
+
+func (g GitChanges) GetVersionBump() string {
+	majorChanges, minorChanges, patchChanges, _ := g.GetSeparatedChanges()
+
+	if len(majorChanges) > 0 {
+		return "MAJOR"
+	} else if len(minorChanges) > 0 {
+		return "MINOR"
+	} else if len(patchChanges) > 0 {
+		return "PATCHES"
+	}
+	return "NONE"
+}
+
+func (g GitChanges) GetSeparatedChanges() ([]GitChange, []GitChange, []GitChange, []GitChange) {
+	majorChanges := make([]GitChange, 0)
+	minorChanges := make([]GitChange, 0)
+	patchChanges := make([]GitChange, 0)
+	unknownChanges := make([]GitChange, 0)
+
+	for _, change := range g {
+		if change.BreakingChange {
+			majorChanges = append(majorChanges, change)
+		} else if _, ok := minorTypes[change.CommitType]; ok {
```

Comment:
If you had a `map[string]string` of featureType->bump, you could combine that with the suggestion above to do this more succinctly:
```
var bump string
var known bool
if bump, known = featureTypes[change.CommitType]; !known {
   bump = "unknown"
} 
bumpMap[bump] = append(bumpMap[bump], change)
```